core/src/test/java/com/linecorp/armeria/server/ServerListenerTest.java

```diff
@@ -100,4 +99,154 @@ public void serverStarting(Server server) throws Exception {
                 .hasCauseInstanceOf(IllegalStateException.class)
                 .hasRootCause(ex);
     }
+
+    @Test
+    void testGracefulShutdownExecutor() throws Exception {
+        final CompletableFuture<Void> stopFuture;
+        final CancellableExecutors executors = new CancellableExecutors();
+        final ServerListener sl = ServerListener
+                .builder()
+                .shutdownWhenStopping(executors.periodicExecutors(0), 0)
+                .shutdownWhenStopping(executors.interruptibleExecutors(0), 0)
+                .shutdownWhenStopping(executors.uninterruptibleExecutors(0), 0)
+                .shutdownWhenStopping(executors.periodicExecutors(1), 1000)
+                .shutdownWhenStopping(executors.interruptibleExecutors(1), 1000)
+                .shutdownWhenStopping(executors.uninterruptibleExecutors(1), 1000)
+                .shutdownWhenStopped(executors.periodicExecutors(2), 0)
+                .shutdownWhenStopped(executors.interruptibleExecutors(2), 0)
+                .shutdownWhenStopped(executors.uninterruptibleExecutors(2), 0)
+                .shutdownWhenStopped(executors.periodicExecutors(3), 1000)
+                .shutdownWhenStopped(executors.interruptibleExecutors(3), 1000)
+                .shutdownWhenStopped(executors.uninterruptibleExecutors(3), 1000)
+                .build();
+
+        final Server server = Server.builder()
+                                    .service("/", (req, ctx) -> HttpResponse.of("Hello!"))
+                                    .serverListener(sl)
+                                    .build();
+        server.start().get();
+        stopFuture = server.stop();
+        Thread.sleep(1000);
+
+        for (int i = 0; i < 2; i++) {
+            assertThat(executors.periodicExecutors(2 * i).isTerminated()).isEqualTo(true);
+            executors.periodicTasks(2 * i).cancel();
+            Thread.sleep(1000);
```

Comment:
Should we use `ArgumentsProvider` to run each case as a separate test instead of running all in a loop?